File: src/MiddleOfLinkedList.java
/* Title: Middle of the Linked List
 * 
 * Problem: Given the head of a singly linked list, return the middle node of the linked list.
 * 
 * Author: Veeral Suthar
 *
 * Date: 07/18/2022
 * 
 * Solution: First we go through the linked list to get the size of the linked list and save its size in the count variable. Then we just go through
 * the the list until we get to the count/2 index and return that value.
 * Time Complexity: O(n)
 */
public class MiddleOfLinkedList {
    public ListNode middleNode(ListNode head) {
        int count = 0;
        ListNode curs = head;
        
        while (curs != null) {
            curs = curs.next;
            count++;
        }
        
        for (int i = 0; i < (count/2);i++) {
            head = head.next;
        }
        return head;
    }
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
}

File: Swift/MiddleOfLinkedList.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class MiddleOfLinkedList {
    func middleNode(_ head: ListNode?) -> ListNode? {
        var middle: Int = llCount(head)/2
        var ret: ListNode? = head
        for i in 0..<middle {
            ret = ret?.next
        }
        return ret
    }

    // Function to get the count of the linked list. Ideally would be a ListNode class function
    func llCount(_ node: ListNode?) -> Int {
        var count: Int = 1
        var pointer: ListNode? = node

        while (pointer?.next != nil) {
            count += 1
            pointer = pointer?.next
        }
        return count
    }
}